src/task_manager/cooperative.rs

```diff
@@ -5,131 +5,257 @@ use crate::task_manager::{
+pub struct CooperativeTask {
+    ///  Contains 3 functions for task execution inherited from the ```Task```: ```setup_fn```,
+    /// ```loop_fn``` and ```stop_condition_fn```.
+    pub(crate) core: Task,
+    /// Each task has a unique ```id```. The First ```id``` number is 0.
+    id: TaskIdType,
+    /// Status of existing ```CooperativeTask```. It may change during the task executing.
+    status: TaskStatusType,
+    /// Each ```CooperativeTask``` has a ```priority```.
+    /// It is taken into account when selecting the next task to execute.
+    priority: TaskPriorityType,
 }
 
-impl Future for FutureTask {
-    type Output = ();
-
-    fn poll(mut self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
-        let mut array: [usize; 8] = core::array::from_fn(|i| i);
-        array[0] = 5;
-        if (self.task.stop_condition_fn)() {
-            Poll::Ready(())
-        } else {
-            if !self.is_setup_completed {
-                (self.task.setup_fn)();
-                self.is_setup_completed = true;
-            } else {
-                (self.task.loop_fn)();
-            }
-            Poll::Pending
+/// Cooperative task manager representation. Based on round-robin scheduling with priorities.
+#[repr(C)]
+pub struct CooperativeTaskManager {
+    /// Array of vectors with ```CooperativeTask``` to execute.
+    pub(crate) tasks: [Vec<CooperativeTask>; NUM_PRIORITIES],
+    /// ```id``` of a task that will be created the next.
+    pub(crate) next_task_id: TaskIdType,
+}
+
+/// Cooperative implementation of ```TaskManagerTrait```.
+impl TaskManagerTrait for CooperativeTaskManager {
+    /// Add a task to task manager.
+    /// It should pass setup, loop, and condition functions.
+    /// Task added with this function has ```priority``` 0.
+    fn add_task(
+        setup_fn: TaskSetupFunctionType,
+        loop_fn: TaskLoopFunctionType,
+        stop_condition_fn: TaskStopConditionFunctionType,
+    ) {
+        CooperativeTaskManager::add_priority_task(setup_fn, loop_fn, stop_condition_fn, 0);
+    }
+
+    /// Starts task manager work.
+    fn start_task_manager() -> ! {
+        loop {
+            CooperativeTaskManager::schedule();
         }
     }
 }
 
-/// Creates simple task waker. May be more difficult in perspective.
-pub fn task_waker() -> Waker {
-    fn raw_clone(_: *const ()) -> RawWaker {
-        RawWaker::new(core::ptr::null::<()>(), &NOOP_WAKER_VTABLE)
+impl CooperativeTaskManager {
+    /// Creates new task manager.
+    pub(crate) const fn new() -> CooperativeTaskManager {
+        let tasks = [
+            Vec::new(),
+            Vec::new(),
+            Vec::new(),
+            Vec::new(),
+            Vec::new(),
+            Vec::new(),
+            Vec::new(),
+            Vec::new(),
+            Vec::new(),
+            Vec::new(),
+            Vec::new(),
+        ];
+        CooperativeTaskManager {
+            tasks,
+            next_task_id: 0,
+        }
```

Comment:
Why can't we begin with empty tasks array?

Reply:
Each vector is for tasks of a specific priority and all possible priorities are initialized at the start of the manager.

Reply:
Why can't we create vector for exact priority, when task with such priority appears? Cause seems like we use unnecessary memory. It may be critical for devices with small memory such as esp32.